### PR TITLE
[libc] Fix linking of AMDGPU device runtime control constants for math

### PR DESCRIPTION
**Pull request by jhuber6**
Summary:
Currently, `libc` temporarily provides math by linking against existing
vendor implementations. To use the AMDGPU DeviceRTL we need to define a
handful of control constants that alter behaviour for architecture
specific things. Previously these were marked `extern const` because
they must be present when we link-in the vendor bitcode library.
However, this causes linker errors if more than one math function was
used.

This patch fixes the issue by marking these functions as used and inline
on top of being external. This means that they are linkable, but it
gives us `linkonce_odr` semantics. The downside is that these globals
won't be optimized out, but it allows us to perform constant propagation
on them unlike using `weak`.

### CODE REVIEW

**Comment by arsenm** [comment]
Is there anything we can do better than used? I don't want these to make it into the final binary

**Comment by jhuber6** [comment]
> Is there anything we can do better than used? I don't want these to make it into the final binary

So, the other option would be to compile this and then `-mlink-builtin-bitcode` it. However the CMake isn't really built for this so it would be a lot of custom hackery.

Another solution would be to emit an internal alias to these variables, that would allow it to be stripped by LTO.

**Comment by jhuber6** [comment]
I updated it to use an internal alias. This causes clang to emit the symbol long enough and with the correct linkage to link against it. The resulting symbols can then be trivially stripped once we optimize the program, see https://godbolt.org/z/66K8T3orW.

**Comment by arsenm** [comment]
Cmake cannot and should not be trying to link these. -mlink-builtin-bitcode is the only correct way to link device libs

**Comment by jhuber6** [comment]
> Cmake cannot and should not be trying to link these. -mlink-builtin-bitcode is the only correct way to link device libs

Sorry, I wasn't specific enough. I was talking about compiling this header as a bitcode file, then using `-mlink-builtin-bitcode` on the device libs. That would be a nightmare of CMake since we'd need to somehow only pass the flag to the correct architecture.

**Comment by JonChesterfield** [comment]
It's sad that the device libs need awkward linking but also very long standing. Either link them awkwardly, don't use them, or link them incorrectly and hope the failure modes don't show up in bug reports. I favour don't use them for libc.

**Comment by jhuber6** [comment]
> It's sad that the device libs need awkward linking but also very long standing. Either link them awkwardly, don't use them, or link them incorrectly and hope the failure modes don't show up in bug reports. I favour don't use them for libc.

We need to use them right now since we're using `ocml.bc` to generate the math libraries. The end goal is to remove this completely however, this is merely a stopgap.

**Comment by arsenm** [comment]
So technically ocml functions are now subtarget independent. They still could have the base wave size and other ABI incompatibilities. The ISA versions are only used for detecting if FMA is fast or not, which barely matters anymore. We could find a way to avoid that

**Comment by jhuber6** [comment]
> So technically ocml functions are now subtarget independent. It's ockl that's still problematic. They still could have the base wave size and other ABI incompatibilities. The ISA versions are only used for detecting if FMA is fast or not, which barely matters anymore. We could find a way to avoid that

Yeah, none of the `ocml` functions use the wavesize. We'd still need to set the math optimization flags though.

**Comment by arsenm** [comment]
The wavesize incompatibility is deeper than depends on the feature. It's a full ABI incompatibility and really should have 2 separate builds.

For the other math options, I have all the patches to eliminate them stuck in review

**Comment by jhuber6** [comment]
> The wavesize incompatibility is deeper than depends on the feature. It's a full ABI incompatibility and really should have 2 separate builds.
> 
> For the other math options, I have all the patches to eliminate them stuck in review

So, this approach is mostly just a quick solution to port the `ocml` libraries to this multi-architecture build I'm doing in `libc`. I think long-term I'd really like to move the builds to be common, since there's generally only a few builtins we could probably just make errors in the backend if they're unsupported once the user provides an architecture. The wavesize attribute is definitely the hardest one, but at least for my purposes we may be able to allocate 64 sized buffers always and use the intrinsic to set the upper limit.

**Comment by jhuber6** [comment]
Ping

File: libc/src/math/gpu/vendor/amdgpu/platform.h [hunk 1]
```diff
@@ -19,96 +21,106 @@ namespace __llvm_libc {
 extern "C" {
 
 // Disable unsafe math optimizations in the implementation.
-extern const uint8_t __oclc_unsafe_math_opt = 0;
+extern const LIBC_INLINE uint8_t __oclc_unsafe_math_opt = 0;
```
Comment (sivachandra) [inline libc/src/math/gpu/vendor/amdgpu/platform.h:24]: `LIBC_INLINE_VAR`

**Comment by jhuber6** [comment]
@arsenm are there any issues with this? It's fixing something that's broken upstream right now without changing any semantics so I'd like to get it in if possible.

**Comment by jhuber6** [comment]
Ping. This is a bug fix that prevents this from actually being used and doesn't change any existing behavior.

**Review by arsenm — APPROVED** [review]

**Pull request merged**